{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "# This program processes FASTA files for analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1",
   "metadata": {},
   "source": [
    "## Parse FASTA File and Match Taxonomic Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import csv\n",
    "import pandas as pd\n",
    "from collections import defaultdict\n",
    "from pathlib import Path\n",
    "import sqlite3\n",
    "from Bio import SeqIO\n",
    "\n",
    "# for debugging\n",
    "from itertools import islice"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3",
   "metadata": {},
   "source": [
    "### Developing Tree Model \"taxo_search_tree\" for Fast Taxonomic Lookup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Table containing all taxonomic relationships from ICTV\n",
    "taxonomy_table = pd.read_csv(\"taxonomy_table.csv\", index_col=0)\n",
    "\n",
    "taxonomy_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract info from taxonomy_table and build lookup tree\n",
    "\n",
    "# define tree\n",
    "taxo_search_tree = defaultdict(str)\n",
    "\n",
    "\n",
    "# iterate through entire table and define relationships\n",
    "\n",
    "# itertuples is faster for row iteration than pandas series\n",
    "for row_tuple in taxonomy_table.itertuples(index=False, name = None):\n",
    "\n",
    "    \n",
    "    # Iterate through levels (EXCLUDE SUBLEVELS) in reverse order (known size of row = 15)\n",
    "    for i in range(14,-2,-2):\n",
    "\n",
    "        raw_value = row_tuple[i]\n",
    "\n",
    "        if pd.notna(raw_value):\n",
    "            current = raw_value.lower()\n",
    "            \n",
    "            if taxo_search_tree[current] == \"\":\n",
    "                # Point lower level to higher\n",
    "\n",
    "                next_level = row_tuple[i-2]\n",
    "\n",
    "                if pd.notna(next_level):\n",
    "                    taxo_search_tree[current] = next_level.lower()\n",
    "                else:\n",
    "                    taxo_search_tree[current] = next_level\n",
    "\n",
    "    # Iterate through sublevels in reverse order\n",
    "    for i in range(13,-1,-2):\n",
    "\n",
    "        \n",
    "        raw_value = row_tuple[i]\n",
    "\n",
    "        if pd.notna(raw_value):\n",
    "            current = raw_value.lower()\n",
    "            \n",
    "            if taxo_search_tree[current] == \"\":\n",
    "                # Point lower level to related level\n",
    "\n",
    "                next_level = row_tuple[i-2]\n",
    "\n",
    "                if pd.notna(next_level):\n",
    "                    taxo_search_tree[current] = next_level.lower()\n",
    "                else:\n",
    "                    taxo_search_tree[current] = next_level\n",
    "\n",
    "# Completed tree is named: taxo_search_tree"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6",
   "metadata": {},
   "source": [
    "### Parse FASTA files into table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Point at folder with fasta files\n",
    "fasta_dir = \"virusDB/NCBI_Protein_Fasta\"\n",
    "\n",
    "# Grab all Fasta files paths\n",
    "\n",
    "fasta_paths = list(Path(fasta_dir).glob(f'*.{\"fsa_aa\"}'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To determine if entry in taxo_search_tree\n",
    "\n",
    "def find_taxo_key(words, taxo_search_tree):\n",
    "    # Checks all consecutive word combinations in the label against the tree keys\n",
    "    # Returns the longest matching key found\n",
    "\n",
    "    N = len(words)\n",
    "    best_match = None\n",
    "    \n",
    "    # 1. Generate Combinations and Search\n",
    "    # Iterate through all possible starting positions (i)\n",
    "    for i in range(N):\n",
    "        # Iterate through all possible ending positions (j)\n",
    "        # Note: We iterate from the start of the combination (i) to the end (N)\n",
    "        for j in range(i, N):\n",
    "            \n",
    "            # Combine the words from index i to j (inclusive)\n",
    "            candidate = \" \".join(words[i : j + 1])\n",
    "            \n",
    "            # 2. O(1) Dictionary Lookup\n",
    "            if candidate in taxo_search_tree:\n",
    "                # Update the best match (we prefer longer matches)\n",
    "                if best_match is None or len(candidate) > len(best_match):\n",
    "                    best_match = candidate\n",
    "                    \n",
    "    return best_match\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_files = []\n",
    "\n",
    "count = 0\n",
    "\n",
    "\n",
    "# islice(iterator, stop) gets items from the iterator until 'stop' is reached.\n",
    "for index, path in enumerate(fasta_paths):\n",
    "    for record in SeqIO.parse(path, \"fasta\"):\n",
    "        full_header = record.description\n",
    "\n",
    "        # extract organism name inside brackets using regex\n",
    "        match = re.search(r'\\[(.+?)\\]', full_header)\n",
    "        organism = match.group(1) if match else \"\"\n",
    "\n",
    "        # Find the longest consecutive substring that is in the tree\n",
    "        taxo_key = find_taxo_key(organism.lower().split(),taxo_search_tree)\n",
    "\n",
    "    \n",
    "        # Check if key in the tree\n",
    "        if taxo_key != None:\n",
    "            count+=1\n",
    "        \n",
    "            \n",
    "print(f\"Count is {count}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create SQLtable for very large table\n",
    "\n",
    "conn = sqlite3.connect(\"bacteria.db\")\n",
    "\n",
    "# Create table scheme with sql\n",
    "conn.execute(\"\"\"\n",
    "CREATE TABLE IF NOT EXISTS sequences (\n",
    "  protein_id TEXT,\n",
    "  protein_seq TEXT,\n",
    "  taxonomic_label TEXT,\n",
    "  description TEXT\n",
    ")\n",
    "\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fasta (path):\n",
    "\n",
    "    # Scan through every record in fasta\n",
    "    rows = []\n",
    "    for record in SeqIO.parse(path, \"fasta\"):\n",
    "\n",
    "        # name of the folder containing fasta\n",
    "        folder_name = os.path.basename(os.path.dirname(path))\n",
    "        \n",
    "        full_header = record.description\n",
    "\n",
    "        # extract organism name inside brackets using regex\n",
    "        match = re.search(r'\\[(.+?)\\]', full_header)\n",
    "        organism = match.group(1) if match else \"\"\n",
    "        \n",
    "        # create substring without bracketed part\n",
    "        no_id = full_header[len(record.id):].lstrip(\" :\")        \n",
    "        pre_description = re.sub(r'\\s*\\[.+?\\]', \"\", no_id)\n",
    "\n",
    "\n",
    "        rows.append({\n",
    "            \"genome_id\": folder_name,\n",
    "            \"protein_id\": record.id,\n",
    "            \"protein_seq\": str(record.seq),\n",
    "            \"taxonomic_label\": organism,    \n",
    "            \"description\": pre_description\n",
    "        })\n",
    "\n",
    "    # print (*rows);\n",
    "    return rows;"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (bacteria-env)",
   "language": "python",
   "name": "bacteria-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
