{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5e98a85f-80e0-441e-9298-9bcf5141d3a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import pandas as pd\n",
    "import ast\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c758fb1b-c0ae-4c7d-b026-c5982308d889",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Processed and wrote a chunk of 5000 rows.\n",
      "Successfully merged all data into mergedTables.csv\n"
     ]
    }
   ],
   "source": [
    "main_file_headers = [\"genome_id\", \"protein_id\", \"protein_seq\", \"taxonomic_label\", \"host_id\"]\n",
    "\n",
    "\n",
    "# Read both csv to be merged\n",
    "df1_iterator = pd.read_csv('/project/jespinoza_1537/GENOMESDB/bacteriaDB/embeddings.csv', chunksize=5000,header=None)\n",
    "df2_iterator = pd.read_csv('/project/jespinoza_1537/GENOMESDB/bacteriaDB/bacteriaRandSub.csv', chunksize=5000,header=None, names=main_file_headers)\n",
    "\n",
    "\n",
    "\n",
    "is_first_chunk = True \n",
    "\n",
    "for chunk1, chunk2 in zip(df1_iterator, df2_iterator):\n",
    "    \n",
    "    # Create copy of chunk 2 and add the embedding to it at the end\n",
    "    merged_chunk = chunk2.copy()\n",
    "\n",
    "    # For every row in out.csv, convert to a list and make it a string\n",
    "    vector_descriptions = chunk1.apply(lambda row: str(row.tolist()), axis=1)\n",
    "\n",
    "    # Did not use join() since can get messy in a loop\n",
    "    merged_chunk['embedding'] = vector_descriptions.values\n",
    "\n",
    "    # Append to mergedTables.csv\n",
    "    if is_first_chunk:\n",
    "            # For the first chunk, write in 'w' mode (overwrite) and include the header.\n",
    "            merged_chunk.to_csv(\"mergedTables.csv\", mode='w', header=True, index=False)\n",
    "            is_first_chunk = False\n",
    "    else:\n",
    "            # For subsequent chunks, append in 'a' mode and omit the header.\n",
    "            merged_chunk.to_csv(\"mergedTables.csv\", mode='a', header=False, index=False)\n",
    "\n",
    "    # Updates if first_chunk so header is not written more then once\n",
    "    is_first_chunk = False\n",
    "\n",
    "    print(f\"Processed and wrote a chunk of {len(chunk1)} rows.\")\n",
    "\n",
    "\n",
    "print(f\"Successfully merged all data into mergedTables.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dfa209f6-6f3a-4cdf-802a-649a4343697b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>protein_seq</th>\n",
       "      <th>taxonomic_label</th>\n",
       "      <th>host_id</th>\n",
       "      <th>embedding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_009811375.1</td>\n",
       "      <td>WP_159277441.1</td>\n",
       "      <td>MQFEEGIAWGVQDGIGRIVLKRPERANSVSLASSHALARAIDEVLE...</td>\n",
       "      <td>Variovorax boronicumulans</td>\n",
       "      <td>enoyl-CoA hydratase/isomerase family protein</td>\n",
       "      <td>[-0.0007136122, -0.15305841, 0.029148007, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_900112575.1</td>\n",
       "      <td>WP_091493282.1</td>\n",
       "      <td>MGIRLKTIRIKGFRGFKNIEVDFQNTTVLVGTNNAGKTTLLKALQV...</td>\n",
       "      <td>Flavobacterium phragmitis</td>\n",
       "      <td>ATP-dependent endonuclease</td>\n",
       "      <td>[0.059893496, 0.029295022, 0.04714787, 0.07690...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_022701015.1</td>\n",
       "      <td>WP_173598177.1</td>\n",
       "      <td>MKVAIMLFDGITALDAIGPYDVFAATLQWEVKFVAKQKGLVKLDSN...</td>\n",
       "      <td>Brevibacillus</td>\n",
       "      <td>MULTISPECIES: DJ-1/PfpI family protein</td>\n",
       "      <td>[0.07244332, -0.079110704, 0.038540285, -0.005...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_029223485.1</td>\n",
       "      <td>WP_275781621.1</td>\n",
       "      <td>MVLTLSAVRWAGEVPKFYRDDCPVPIFAPNSMPTQEDNKVATALGF...</td>\n",
       "      <td>Streptomyces coacervatus</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[-0.05294113, 0.07054596, 0.046061877, -0.0093...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_039532985.1</td>\n",
       "      <td>WP_344363376.1</td>\n",
       "      <td>MGTGTEGFAEALRGLKERSGLSYGALAKRAHMSTSTLHRYCNGDAV...</td>\n",
       "      <td>Streptomyces gobitricini</td>\n",
       "      <td>helix-turn-helix domain-containing protein</td>\n",
       "      <td>[0.0003043313, -0.09910503, -0.034521643, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GCF_042846945.1</td>\n",
       "      <td>WP_390955654.1</td>\n",
       "      <td>MPSIIFFAWALAAITVVVHAIGLALILKALARLETLHIHRLWAIIN...</td>\n",
       "      <td>Pseudomonas moorei</td>\n",
       "      <td>ion channel</td>\n",
       "      <td>[0.004557234, -0.07350842, -0.015117904, -0.01...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>GCF_009883735.1</td>\n",
       "      <td>WP_159406565.1</td>\n",
       "      <td>MASRSDLDGKRVLVSGASGFTGRYLTQELREQGCTVIGVGTRAESA...</td>\n",
       "      <td>Xanthomonas</td>\n",
       "      <td>MULTISPECIES: GDP-mannose 4,6-dehydratase</td>\n",
       "      <td>[-0.012554529, -0.022618555, 0.048023347, 0.01...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>GCF_042648425.1</td>\n",
       "      <td>WP_377743539.1</td>\n",
       "      <td>MIRRIPHCSHWGAYTILVEDGRILGVEPFAGDPAPSPVIHSVAEWA...</td>\n",
       "      <td>Paracoccus mangrovi</td>\n",
       "      <td>molybdopterin-dependent oxidoreductase</td>\n",
       "      <td>[-0.025443427, -0.01865664, -0.036623105, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>GCF_003318295.1</td>\n",
       "      <td>WP_113805433.1</td>\n",
       "      <td>MKIAFLSDIHGNADALEAVIRDIEAKQVDKIVVLGDICYRGPEPKR...</td>\n",
       "      <td>Bacillus taeanensis</td>\n",
       "      <td>metallophosphoesterase family protein</td>\n",
       "      <td>[0.0847185, -0.08239446, 0.0056341896, 0.04096...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>GCF_031625015.1</td>\n",
       "      <td>WP_310890600.1</td>\n",
       "      <td>MSRRLADGIQAEPPRSDRPVETAIVALGANLGDRVETLEAAVADIR...</td>\n",
       "      <td>Microbacterium aurantiacum</td>\n",
       "      <td>2-amino-4-hydroxy-6-hydroxymethyldihydropterid...</td>\n",
       "      <td>[-0.0405488, -0.046200313, 0.03950226, -0.0107...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>GCF_003176835.1</td>\n",
       "      <td>WP_162551474.1</td>\n",
       "      <td>MEKPKNGFSIRKLGVTLCLTIIVILGLIFTFQSPLTLWMLGHHEDK...</td>\n",
       "      <td>Lactococcus lactis</td>\n",
       "      <td>class A sortase</td>\n",
       "      <td>[-0.01416887, 0.07933725, 0.036645338, -0.0026...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>GCF_014217765.1</td>\n",
       "      <td>WP_072735992.1</td>\n",
       "      <td>MSKFRVDLDRLDETISKVETFDSNVDDIITAVEGHLADLHETWTGR...</td>\n",
       "      <td>Rhodococcus triatomae</td>\n",
       "      <td>WXG100 family type VII secretion target</td>\n",
       "      <td>[-0.12231629, -0.045867145, -0.10942466, 0.034...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>GCF_000420785.1</td>\n",
       "      <td>WP_281167107.1</td>\n",
       "      <td>MTVKQTLTKAEKEFLETYHSRSYHYFKDIKKYVALLEKLTKV</td>\n",
       "      <td>Streptococcus hyovaginalis</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[0.06882511, -0.014589515, -0.052757192, 0.006...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>GCF_003627075.1</td>\n",
       "      <td>WP_157981639.1</td>\n",
       "      <td>MADRAAGSKPTPRRSTAGQAGKAAVSIAAAFALTATGGMPAVAAPR...</td>\n",
       "      <td>Protaetiibacter intestinalis</td>\n",
       "      <td>NlpC/P60 family protein</td>\n",
       "      <td>[-0.04632308, 0.048107393, -0.028366879, 0.059...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>GCF_029223485.1</td>\n",
       "      <td>WP_275780080.1</td>\n",
       "      <td>MSGAAVVAVSVEVIVFFLVLSPVSLAVVSSIVLAEAVRSVVSFKAL...</td>\n",
       "      <td>Streptomyces coacervatus</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[0.012212416, -0.08464475, 0.14822036, -0.0195...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>GCF_022846515.1</td>\n",
       "      <td>WP_081118266.1</td>\n",
       "      <td>MTQRTEKEKMIAGELYFSPDPELVADRKYAREQMNLINQETDTTIK...</td>\n",
       "      <td>Enterococcus</td>\n",
       "      <td>MULTISPECIES: chorismate mutase</td>\n",
       "      <td>[0.051910114, 0.036717046, 0.029067008, -0.001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GCF_002993285.1</td>\n",
       "      <td>WP_199832534.1</td>\n",
       "      <td>MKLRLIAATAVGAAAAVTAGLLAPHAAAEPASPAHQPVKIAPAPAV...</td>\n",
       "      <td>Gordonia iterans</td>\n",
       "      <td>lipase family protein</td>\n",
       "      <td>[-0.045864064, 0.015817003, -0.042270146, 0.05...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GCF_015207485.1</td>\n",
       "      <td>WP_193871155.1</td>\n",
       "      <td>MDVNAFQPVFDMLGLFKSRLIFLLIALGVFAAFDFKVFLGSLFMVA...</td>\n",
       "      <td>Planktothrix mougeotii</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[0.05457813, -0.061550744, -0.0016159908, 0.00...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>GCF_001700515.1</td>\n",
       "      <td>WP_066956943.1</td>\n",
       "      <td>MDLRLNGKRALVTGSSTGLGEEIAKVLAAEGAAVVVHGRDEARATA...</td>\n",
       "      <td>Streptomyces lushanensis</td>\n",
       "      <td>SDR family NAD(P)-dependent oxidoreductase, pa...</td>\n",
       "      <td>[-0.00037843146, -0.031833712, 0.022331744, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>GCF_009600885.1</td>\n",
       "      <td>WP_194292333.1</td>\n",
       "      <td>MASVQDEIPGLVIYAVLQPDHDESSIQARFEEFHRLNPWVLAALEK...</td>\n",
       "      <td>Streptomyces jumonjinensis</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[0.046004813, 0.026324565, 0.00943889, -0.0344...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>GCF_001278705.1</td>\n",
       "      <td>WP_053604665.1</td>\n",
       "      <td>MSKKPAYEPLPIQFYNKPTLELAKALLGCLLVKETDEGICSGYIVE...</td>\n",
       "      <td>Bacillus</td>\n",
       "      <td>MULTISPECIES: DNA-3-methyladenine glycosylase</td>\n",
       "      <td>[0.060048643, -0.020489886, 0.003128626, -0.01...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>GCF_045041415.1</td>\n",
       "      <td>WP_164385105.1</td>\n",
       "      <td>MSRKPELKGRAARLARLLPAGSARGQAARAPFVLLVVVLLGGGLIG...</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>MULTISPECIES: septum formation initiator famil...</td>\n",
       "      <td>[-0.03346974, -0.07286795, -0.13021637, -0.007...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>GCF_042984175.1</td>\n",
       "      <td>WP_391591560.1</td>\n",
       "      <td>MQNFKTNILILFFIFLSALIGSGFLLKIKKYEVNYLQNFQTFQKLN...</td>\n",
       "      <td>[Mycoplasma</td>\n",
       "      <td>hypothetical protein cavipharyngis]</td>\n",
       "      <td>[0.09801802, 0.020838303, 0.035793103, 0.05458...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>GCF_006334995.2</td>\n",
       "      <td>WP_226961349.1</td>\n",
       "      <td>MPHPPNTVERPKRLLTQNRRLRSMGIWTWTLPAWAGRFPDGRTYNT...</td>\n",
       "      <td>Streptomyces</td>\n",
       "      <td>MULTISPECIES: GP88 family protein</td>\n",
       "      <td>[-0.071274094, -0.07724204, -0.054615665, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>GCF_008693185.1</td>\n",
       "      <td>WP_070068752.1</td>\n",
       "      <td>MISTTRQYNVISTGISVCTLCFITTSTYAVDLNSQDLIPAPAGTNA...</td>\n",
       "      <td>Acinetobacter qingfengensis</td>\n",
       "      <td>transporter</td>\n",
       "      <td>[-0.06612152, -0.05528904, -0.0983323, 0.01714...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>GCF_007750655.1</td>\n",
       "      <td>WP_145181880.1</td>\n",
       "      <td>MPHQSELELERIVRRLLPILWPAGFEFELSELGVNSGGSFAAGFFS...</td>\n",
       "      <td>Planctomycetes bacterium Pla163</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[-0.0019091254, 0.014995057, 0.04013634, -0.02...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>GCF_002953735.1</td>\n",
       "      <td>WP_104967695.1</td>\n",
       "      <td>MFSLSQISTQFSSFSAGSDFTDFVKTTPDWAKTVNSEWEKRTKVND...</td>\n",
       "      <td>Streptococcus pluranimalium</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[0.018412575, 0.014546668, -0.023422927, 0.094...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>GCF_042651685.1</td>\n",
       "      <td>WP_307230962.1</td>\n",
       "      <td>MALPNLETKESKGLTRFLNGIERIGNKLPDPFIIFVYLMLIVIIVS...</td>\n",
       "      <td>Bacillus chungangensis</td>\n",
       "      <td>AbgT family transporter</td>\n",
       "      <td>[0.042410787, -0.049514998, -0.061050948, 0.04...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>GCF_000367945.1</td>\n",
       "      <td>WP_004656714.1</td>\n",
       "      <td>MILIDSVYINDGGGLVLLKYLVDVLGKTELSIFYLFDERTKDIFKD...</td>\n",
       "      <td>Acinetobacter proteolyticus</td>\n",
       "      <td>glycosyltransferase</td>\n",
       "      <td>[-0.017061016, -0.087328464, -0.019075712, 0.1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>GCF_002749895.1</td>\n",
       "      <td>WP_245869190.1</td>\n",
       "      <td>PCALVISTPAAITSGLAAGARRGVLIKGGAALEQLGRIQVMAFDKT...</td>\n",
       "      <td>Vibrio fujianensis</td>\n",
       "      <td>HAD family hydrolase, partial</td>\n",
       "      <td>[0.022794552, -0.11553865, 0.012081963, 0.0831...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>GCF_016909125.1</td>\n",
       "      <td>WP_204653103.1</td>\n",
       "      <td>MFESKEVYQLSQSKKMMVFVLSMSLYGLATLFTELIPSIHLGVIEF...</td>\n",
       "      <td>Enterococcus lemanii</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[0.008736651, -0.0027261714, 0.015496009, 0.02...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>GCF_042665355.1</td>\n",
       "      <td>WP_381834606.1</td>\n",
       "      <td>MDLTVFGATGGIGQEVVRQALDAGHRVTAVVREPARLPVTGERLRV...</td>\n",
       "      <td>Streptomyces monticola</td>\n",
       "      <td>NAD(P)-dependent oxidoreductase</td>\n",
       "      <td>[0.009260033, -0.036371656, 0.05298036, -0.055...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>GCF_013388375.1</td>\n",
       "      <td>WP_012437158.1</td>\n",
       "      <td>MVAMMPLHATTARLERSTMIPTSLKKSVRPRDARGRFIRHHERLPV...</td>\n",
       "      <td>Xanthomonas</td>\n",
       "      <td>MULTISPECIES: glycogen synthase GlgA</td>\n",
       "      <td>[0.002926987, 0.010527462, -0.03377514, -0.068...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>GCF_016919245.1</td>\n",
       "      <td>WP_205372179.1</td>\n",
       "      <td>MDRHHLTTTRRGSGTANDPTWTGWIYAALGALAAAQLLSGLVGAGA...</td>\n",
       "      <td>Streptomyces zhihengii</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[-0.039358903, 0.015897823, -0.035706297, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>GCF_003851555.1</td>\n",
       "      <td>WP_164484823.1</td>\n",
       "      <td>MFSHEGTALELLILVFGVTGVLLIAWFFHHQYELRARKRVKKQHSP...</td>\n",
       "      <td>Pseudomonas synxantha</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[-0.028012691, -0.038667988, -0.022580912, -0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>GCF_000022005.1</td>\n",
       "      <td>YP_002518661.1</td>\n",
       "      <td>MADYIALIHKDADSAFGVSFPDLPGCISVGETLEEARANAEEALAL...</td>\n",
       "      <td>Caulobacter vibrioides NA1000</td>\n",
       "      <td>type II toxin-antitoxin system HicB family ant...</td>\n",
       "      <td>[-0.009591965, -0.0054962244, -0.008626771, -0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>GCF_014337925.2</td>\n",
       "      <td>WP_187428932.1</td>\n",
       "      <td>MDVNHAPKKLLNAGSHRVCSVEETLARILPMRHEFGITRIANVTGL...</td>\n",
       "      <td>Roseobacter fucihabitans</td>\n",
       "      <td>YcaO-like family protein</td>\n",
       "      <td>[-0.047342326, 0.03945248, 0.030565826, -0.049...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>GCF_004770155.1</td>\n",
       "      <td>WP_135594418.1</td>\n",
       "      <td>MSTKNALIIGSSGVAGQSAIQAVREHSKKRNETWRIVSTTSKNVPL...</td>\n",
       "      <td>Leptospira gomenensis</td>\n",
       "      <td>SDR family NAD(P)-dependent oxidoreductase</td>\n",
       "      <td>[0.03776277, 0.026553273, 0.022663444, -0.0016...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>GCF_004522175.1</td>\n",
       "      <td>WP_134682358.1</td>\n",
       "      <td>MSSEEDSTAKWIEGLQGADALFVSQVRLKPEFVLQALIGAREQGSA...</td>\n",
       "      <td>Paracoccus ravus</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[0.040732816, 0.004375422, -0.02246262, -0.014...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>GCF_017883985.1</td>\n",
       "      <td>WP_209627744.1</td>\n",
       "      <td>MKILITSGGTSEKIDRVRSITNHSTGQLGKLIAETFLDRGHQVTLV...</td>\n",
       "      <td>Streptococcus oricebi</td>\n",
       "      <td>phosphopantothenate--cysteine ligase</td>\n",
       "      <td>[0.06212287, -0.119477965, 0.002100817, 0.0573...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>GCF_029872515.1</td>\n",
       "      <td>WP_280309448.1</td>\n",
       "      <td>MSRTLAIFAVACSVLVCTPAAAFGLKTHLWIGQQIIAEVNQSCRIR...</td>\n",
       "      <td>Pseudomonas flavocrustae</td>\n",
       "      <td>zinc dependent phospholipase C family protein</td>\n",
       "      <td>[-0.035855744, -0.0048338864, -0.008274413, 0....</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>GCF_039531105.1</td>\n",
       "      <td>WP_344053657.1</td>\n",
       "      <td>MSFSLPADPPRSRSLTGVVPQMLAAIGGSSDWFAPARSAIVFVVDG...</td>\n",
       "      <td>Microbacterium lacus</td>\n",
       "      <td>nucleotide pyrophosphatase/phosphodiesterase f...</td>\n",
       "      <td>[-0.00273796, -0.013839726, 0.014343991, -0.03...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>GCF_001314995.1</td>\n",
       "      <td>WP_004325012.1</td>\n",
       "      <td>MKKTIYTAIATLLFLISSCTLEQEVYNKINPGMFPQNAEDVNALVN...</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>MULTISPECIES: RagB/SusD family nutrient uptake...</td>\n",
       "      <td>[-0.055298418, 0.01586196, -0.011672523, 0.104...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>GCF_007988825.1</td>\n",
       "      <td>WP_147039462.1</td>\n",
       "      <td>MLLTESVEVPVSRRNARGRWRRFTAWRRRRPFLGGLLIAVAGVEMF...</td>\n",
       "      <td>Microbacterium aerolatum</td>\n",
       "      <td>DUF6114 domain-containing protein</td>\n",
       "      <td>[-0.047459356, 0.041661527, -0.021046067, -0.0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>GCF_904830955.1</td>\n",
       "      <td>WP_254175435.1</td>\n",
       "      <td>MKTIKNSVTTNNKTQKPSADIRSNTLFTDITAQEEVFVRGGMSRIAGAS</td>\n",
       "      <td>Planktothrix tepida</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[-0.060417198, 0.0048286803, -0.03349639, 0.11...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>GCF_025822065.1</td>\n",
       "      <td>WP_069419754.1</td>\n",
       "      <td>MAQPRLDDDRAAEGGNPFPVDSTTRPCCSGIGNHTRDCRGDLLEES...</td>\n",
       "      <td>Mycobacterium intermedium</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[0.023872152, 0.008181557, 0.032098502, -0.001...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>GCF_004564075.1</td>\n",
       "      <td>WP_067028234.1</td>\n",
       "      <td>MPLTPADVEFSVDQSRIDQLWVHETLQTHAYWVEGRTAEMQLRANQ...</td>\n",
       "      <td>Microbacterium sediminis</td>\n",
       "      <td>GNAT family N-acetyltransferase</td>\n",
       "      <td>[0.042925302, -0.009820527, -0.02448054, -0.02...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>GCF_016064815.1</td>\n",
       "      <td>WP_198072607.1</td>\n",
       "      <td>MFSISNNELYSGKSSPLFNKEIASKVRNLHRKIEGYQPTPLISLEY...</td>\n",
       "      <td>Acinetobacter calcoaceticus/baumannii complex</td>\n",
       "      <td>MULTISPECIES: diaminopropionate ammonia-lyase</td>\n",
       "      <td>[0.04361059, -0.060083244, 0.026317095, 0.0907...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>GCF_001277215.2</td>\n",
       "      <td>WP_023899257.1</td>\n",
       "      <td>MLNAWHLPVAPFIHRQGEKLLITLWLAGDALPEKVMLRYERDNEET...</td>\n",
       "      <td>Cronobacter malonaticus</td>\n",
       "      <td>maltodextrin glucosidase</td>\n",
       "      <td>[0.018182483, -0.08957952, 0.048921805, -0.066...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>GCF_020819595.1</td>\n",
       "      <td>WP_228619132.1</td>\n",
       "      <td>MSPGTGEVYRIDWLPGTDLLHGTCHCGAQHTAEDPVLMWEWMLAHP...</td>\n",
       "      <td>Streptomyces barringtoniae</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>[-0.022645863, 0.0143042, -8.128404e-05, -0.06...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          genome_id      protein_id  \\\n",
       "0   GCF_009811375.1  WP_159277441.1   \n",
       "1   GCF_900112575.1  WP_091493282.1   \n",
       "2   GCF_022701015.1  WP_173598177.1   \n",
       "3   GCF_029223485.1  WP_275781621.1   \n",
       "4   GCF_039532985.1  WP_344363376.1   \n",
       "5   GCF_042846945.1  WP_390955654.1   \n",
       "6   GCF_009883735.1  WP_159406565.1   \n",
       "7   GCF_042648425.1  WP_377743539.1   \n",
       "8   GCF_003318295.1  WP_113805433.1   \n",
       "9   GCF_031625015.1  WP_310890600.1   \n",
       "10  GCF_003176835.1  WP_162551474.1   \n",
       "11  GCF_014217765.1  WP_072735992.1   \n",
       "12  GCF_000420785.1  WP_281167107.1   \n",
       "13  GCF_003627075.1  WP_157981639.1   \n",
       "14  GCF_029223485.1  WP_275780080.1   \n",
       "15  GCF_022846515.1  WP_081118266.1   \n",
       "16  GCF_002993285.1  WP_199832534.1   \n",
       "17  GCF_015207485.1  WP_193871155.1   \n",
       "18  GCF_001700515.1  WP_066956943.1   \n",
       "19  GCF_009600885.1  WP_194292333.1   \n",
       "20  GCF_001278705.1  WP_053604665.1   \n",
       "21  GCF_045041415.1  WP_164385105.1   \n",
       "22  GCF_042984175.1  WP_391591560.1   \n",
       "23  GCF_006334995.2  WP_226961349.1   \n",
       "24  GCF_008693185.1  WP_070068752.1   \n",
       "25  GCF_007750655.1  WP_145181880.1   \n",
       "26  GCF_002953735.1  WP_104967695.1   \n",
       "27  GCF_042651685.1  WP_307230962.1   \n",
       "28  GCF_000367945.1  WP_004656714.1   \n",
       "29  GCF_002749895.1  WP_245869190.1   \n",
       "30  GCF_016909125.1  WP_204653103.1   \n",
       "31  GCF_042665355.1  WP_381834606.1   \n",
       "32  GCF_013388375.1  WP_012437158.1   \n",
       "33  GCF_016919245.1  WP_205372179.1   \n",
       "34  GCF_003851555.1  WP_164484823.1   \n",
       "35  GCF_000022005.1  YP_002518661.1   \n",
       "36  GCF_014337925.2  WP_187428932.1   \n",
       "37  GCF_004770155.1  WP_135594418.1   \n",
       "38  GCF_004522175.1  WP_134682358.1   \n",
       "39  GCF_017883985.1  WP_209627744.1   \n",
       "40  GCF_029872515.1  WP_280309448.1   \n",
       "41  GCF_039531105.1  WP_344053657.1   \n",
       "42  GCF_001314995.1  WP_004325012.1   \n",
       "43  GCF_007988825.1  WP_147039462.1   \n",
       "44  GCF_904830955.1  WP_254175435.1   \n",
       "45  GCF_025822065.1  WP_069419754.1   \n",
       "46  GCF_004564075.1  WP_067028234.1   \n",
       "47  GCF_016064815.1  WP_198072607.1   \n",
       "48  GCF_001277215.2  WP_023899257.1   \n",
       "49  GCF_020819595.1  WP_228619132.1   \n",
       "\n",
       "                                          protein_seq  \\\n",
       "0   MQFEEGIAWGVQDGIGRIVLKRPERANSVSLASSHALARAIDEVLE...   \n",
       "1   MGIRLKTIRIKGFRGFKNIEVDFQNTTVLVGTNNAGKTTLLKALQV...   \n",
       "2   MKVAIMLFDGITALDAIGPYDVFAATLQWEVKFVAKQKGLVKLDSN...   \n",
       "3   MVLTLSAVRWAGEVPKFYRDDCPVPIFAPNSMPTQEDNKVATALGF...   \n",
       "4   MGTGTEGFAEALRGLKERSGLSYGALAKRAHMSTSTLHRYCNGDAV...   \n",
       "5   MPSIIFFAWALAAITVVVHAIGLALILKALARLETLHIHRLWAIIN...   \n",
       "6   MASRSDLDGKRVLVSGASGFTGRYLTQELREQGCTVIGVGTRAESA...   \n",
       "7   MIRRIPHCSHWGAYTILVEDGRILGVEPFAGDPAPSPVIHSVAEWA...   \n",
       "8   MKIAFLSDIHGNADALEAVIRDIEAKQVDKIVVLGDICYRGPEPKR...   \n",
       "9   MSRRLADGIQAEPPRSDRPVETAIVALGANLGDRVETLEAAVADIR...   \n",
       "10  MEKPKNGFSIRKLGVTLCLTIIVILGLIFTFQSPLTLWMLGHHEDK...   \n",
       "11  MSKFRVDLDRLDETISKVETFDSNVDDIITAVEGHLADLHETWTGR...   \n",
       "12         MTVKQTLTKAEKEFLETYHSRSYHYFKDIKKYVALLEKLTKV   \n",
       "13  MADRAAGSKPTPRRSTAGQAGKAAVSIAAAFALTATGGMPAVAAPR...   \n",
       "14  MSGAAVVAVSVEVIVFFLVLSPVSLAVVSSIVLAEAVRSVVSFKAL...   \n",
       "15  MTQRTEKEKMIAGELYFSPDPELVADRKYAREQMNLINQETDTTIK...   \n",
       "16  MKLRLIAATAVGAAAAVTAGLLAPHAAAEPASPAHQPVKIAPAPAV...   \n",
       "17  MDVNAFQPVFDMLGLFKSRLIFLLIALGVFAAFDFKVFLGSLFMVA...   \n",
       "18  MDLRLNGKRALVTGSSTGLGEEIAKVLAAEGAAVVVHGRDEARATA...   \n",
       "19  MASVQDEIPGLVIYAVLQPDHDESSIQARFEEFHRLNPWVLAALEK...   \n",
       "20  MSKKPAYEPLPIQFYNKPTLELAKALLGCLLVKETDEGICSGYIVE...   \n",
       "21  MSRKPELKGRAARLARLLPAGSARGQAARAPFVLLVVVLLGGGLIG...   \n",
       "22  MQNFKTNILILFFIFLSALIGSGFLLKIKKYEVNYLQNFQTFQKLN...   \n",
       "23  MPHPPNTVERPKRLLTQNRRLRSMGIWTWTLPAWAGRFPDGRTYNT...   \n",
       "24  MISTTRQYNVISTGISVCTLCFITTSTYAVDLNSQDLIPAPAGTNA...   \n",
       "25  MPHQSELELERIVRRLLPILWPAGFEFELSELGVNSGGSFAAGFFS...   \n",
       "26  MFSLSQISTQFSSFSAGSDFTDFVKTTPDWAKTVNSEWEKRTKVND...   \n",
       "27  MALPNLETKESKGLTRFLNGIERIGNKLPDPFIIFVYLMLIVIIVS...   \n",
       "28  MILIDSVYINDGGGLVLLKYLVDVLGKTELSIFYLFDERTKDIFKD...   \n",
       "29  PCALVISTPAAITSGLAAGARRGVLIKGGAALEQLGRIQVMAFDKT...   \n",
       "30  MFESKEVYQLSQSKKMMVFVLSMSLYGLATLFTELIPSIHLGVIEF...   \n",
       "31  MDLTVFGATGGIGQEVVRQALDAGHRVTAVVREPARLPVTGERLRV...   \n",
       "32  MVAMMPLHATTARLERSTMIPTSLKKSVRPRDARGRFIRHHERLPV...   \n",
       "33  MDRHHLTTTRRGSGTANDPTWTGWIYAALGALAAAQLLSGLVGAGA...   \n",
       "34  MFSHEGTALELLILVFGVTGVLLIAWFFHHQYELRARKRVKKQHSP...   \n",
       "35  MADYIALIHKDADSAFGVSFPDLPGCISVGETLEEARANAEEALAL...   \n",
       "36  MDVNHAPKKLLNAGSHRVCSVEETLARILPMRHEFGITRIANVTGL...   \n",
       "37  MSTKNALIIGSSGVAGQSAIQAVREHSKKRNETWRIVSTTSKNVPL...   \n",
       "38  MSSEEDSTAKWIEGLQGADALFVSQVRLKPEFVLQALIGAREQGSA...   \n",
       "39  MKILITSGGTSEKIDRVRSITNHSTGQLGKLIAETFLDRGHQVTLV...   \n",
       "40  MSRTLAIFAVACSVLVCTPAAAFGLKTHLWIGQQIIAEVNQSCRIR...   \n",
       "41  MSFSLPADPPRSRSLTGVVPQMLAAIGGSSDWFAPARSAIVFVVDG...   \n",
       "42  MKKTIYTAIATLLFLISSCTLEQEVYNKINPGMFPQNAEDVNALVN...   \n",
       "43  MLLTESVEVPVSRRNARGRWRRFTAWRRRRPFLGGLLIAVAGVEMF...   \n",
       "44  MKTIKNSVTTNNKTQKPSADIRSNTLFTDITAQEEVFVRGGMSRIAGAS   \n",
       "45  MAQPRLDDDRAAEGGNPFPVDSTTRPCCSGIGNHTRDCRGDLLEES...   \n",
       "46  MPLTPADVEFSVDQSRIDQLWVHETLQTHAYWVEGRTAEMQLRANQ...   \n",
       "47  MFSISNNELYSGKSSPLFNKEIASKVRNLHRKIEGYQPTPLISLEY...   \n",
       "48  MLNAWHLPVAPFIHRQGEKLLITLWLAGDALPEKVMLRYERDNEET...   \n",
       "49  MSPGTGEVYRIDWLPGTDLLHGTCHCGAQHTAEDPVLMWEWMLAHP...   \n",
       "\n",
       "                                  taxonomic_label  \\\n",
       "0                       Variovorax boronicumulans   \n",
       "1                       Flavobacterium phragmitis   \n",
       "2                                   Brevibacillus   \n",
       "3                        Streptomyces coacervatus   \n",
       "4                        Streptomyces gobitricini   \n",
       "5                              Pseudomonas moorei   \n",
       "6                                     Xanthomonas   \n",
       "7                             Paracoccus mangrovi   \n",
       "8                             Bacillus taeanensis   \n",
       "9                      Microbacterium aurantiacum   \n",
       "10                             Lactococcus lactis   \n",
       "11                          Rhodococcus triatomae   \n",
       "12                     Streptococcus hyovaginalis   \n",
       "13                   Protaetiibacter intestinalis   \n",
       "14                       Streptomyces coacervatus   \n",
       "15                                   Enterococcus   \n",
       "16                               Gordonia iterans   \n",
       "17                         Planktothrix mougeotii   \n",
       "18                       Streptomyces lushanensis   \n",
       "19                     Streptomyces jumonjinensis   \n",
       "20                                       Bacillus   \n",
       "21                                   Streptomyces   \n",
       "22                                    [Mycoplasma   \n",
       "23                                   Streptomyces   \n",
       "24                    Acinetobacter qingfengensis   \n",
       "25                Planctomycetes bacterium Pla163   \n",
       "26                    Streptococcus pluranimalium   \n",
       "27                         Bacillus chungangensis   \n",
       "28                    Acinetobacter proteolyticus   \n",
       "29                             Vibrio fujianensis   \n",
       "30                           Enterococcus lemanii   \n",
       "31                         Streptomyces monticola   \n",
       "32                                    Xanthomonas   \n",
       "33                         Streptomyces zhihengii   \n",
       "34                          Pseudomonas synxantha   \n",
       "35                  Caulobacter vibrioides NA1000   \n",
       "36                       Roseobacter fucihabitans   \n",
       "37                          Leptospira gomenensis   \n",
       "38                               Paracoccus ravus   \n",
       "39                          Streptococcus oricebi   \n",
       "40                       Pseudomonas flavocrustae   \n",
       "41                           Microbacterium lacus   \n",
       "42                                    Bacteroides   \n",
       "43                       Microbacterium aerolatum   \n",
       "44                            Planktothrix tepida   \n",
       "45                      Mycobacterium intermedium   \n",
       "46                       Microbacterium sediminis   \n",
       "47  Acinetobacter calcoaceticus/baumannii complex   \n",
       "48                        Cronobacter malonaticus   \n",
       "49                     Streptomyces barringtoniae   \n",
       "\n",
       "                                              host_id  \\\n",
       "0        enoyl-CoA hydratase/isomerase family protein   \n",
       "1                          ATP-dependent endonuclease   \n",
       "2              MULTISPECIES: DJ-1/PfpI family protein   \n",
       "3                                hypothetical protein   \n",
       "4          helix-turn-helix domain-containing protein   \n",
       "5                                         ion channel   \n",
       "6           MULTISPECIES: GDP-mannose 4,6-dehydratase   \n",
       "7              molybdopterin-dependent oxidoreductase   \n",
       "8               metallophosphoesterase family protein   \n",
       "9   2-amino-4-hydroxy-6-hydroxymethyldihydropterid...   \n",
       "10                                    class A sortase   \n",
       "11            WXG100 family type VII secretion target   \n",
       "12                               hypothetical protein   \n",
       "13                            NlpC/P60 family protein   \n",
       "14                               hypothetical protein   \n",
       "15                    MULTISPECIES: chorismate mutase   \n",
       "16                              lipase family protein   \n",
       "17                               hypothetical protein   \n",
       "18  SDR family NAD(P)-dependent oxidoreductase, pa...   \n",
       "19                               hypothetical protein   \n",
       "20      MULTISPECIES: DNA-3-methyladenine glycosylase   \n",
       "21  MULTISPECIES: septum formation initiator famil...   \n",
       "22                hypothetical protein cavipharyngis]   \n",
       "23                  MULTISPECIES: GP88 family protein   \n",
       "24                                        transporter   \n",
       "25                               hypothetical protein   \n",
       "26                               hypothetical protein   \n",
       "27                            AbgT family transporter   \n",
       "28                                glycosyltransferase   \n",
       "29                      HAD family hydrolase, partial   \n",
       "30                               hypothetical protein   \n",
       "31                    NAD(P)-dependent oxidoreductase   \n",
       "32               MULTISPECIES: glycogen synthase GlgA   \n",
       "33                               hypothetical protein   \n",
       "34                               hypothetical protein   \n",
       "35  type II toxin-antitoxin system HicB family ant...   \n",
       "36                           YcaO-like family protein   \n",
       "37         SDR family NAD(P)-dependent oxidoreductase   \n",
       "38                               hypothetical protein   \n",
       "39               phosphopantothenate--cysteine ligase   \n",
       "40      zinc dependent phospholipase C family protein   \n",
       "41  nucleotide pyrophosphatase/phosphodiesterase f...   \n",
       "42  MULTISPECIES: RagB/SusD family nutrient uptake...   \n",
       "43                  DUF6114 domain-containing protein   \n",
       "44                               hypothetical protein   \n",
       "45                               hypothetical protein   \n",
       "46                    GNAT family N-acetyltransferase   \n",
       "47      MULTISPECIES: diaminopropionate ammonia-lyase   \n",
       "48                           maltodextrin glucosidase   \n",
       "49                               hypothetical protein   \n",
       "\n",
       "                                            embedding  \n",
       "0   [-0.0007136122, -0.15305841, 0.029148007, -0.0...  \n",
       "1   [0.059893496, 0.029295022, 0.04714787, 0.07690...  \n",
       "2   [0.07244332, -0.079110704, 0.038540285, -0.005...  \n",
       "3   [-0.05294113, 0.07054596, 0.046061877, -0.0093...  \n",
       "4   [0.0003043313, -0.09910503, -0.034521643, -0.0...  \n",
       "5   [0.004557234, -0.07350842, -0.015117904, -0.01...  \n",
       "6   [-0.012554529, -0.022618555, 0.048023347, 0.01...  \n",
       "7   [-0.025443427, -0.01865664, -0.036623105, -0.0...  \n",
       "8   [0.0847185, -0.08239446, 0.0056341896, 0.04096...  \n",
       "9   [-0.0405488, -0.046200313, 0.03950226, -0.0107...  \n",
       "10  [-0.01416887, 0.07933725, 0.036645338, -0.0026...  \n",
       "11  [-0.12231629, -0.045867145, -0.10942466, 0.034...  \n",
       "12  [0.06882511, -0.014589515, -0.052757192, 0.006...  \n",
       "13  [-0.04632308, 0.048107393, -0.028366879, 0.059...  \n",
       "14  [0.012212416, -0.08464475, 0.14822036, -0.0195...  \n",
       "15  [0.051910114, 0.036717046, 0.029067008, -0.001...  \n",
       "16  [-0.045864064, 0.015817003, -0.042270146, 0.05...  \n",
       "17  [0.05457813, -0.061550744, -0.0016159908, 0.00...  \n",
       "18  [-0.00037843146, -0.031833712, 0.022331744, 0....  \n",
       "19  [0.046004813, 0.026324565, 0.00943889, -0.0344...  \n",
       "20  [0.060048643, -0.020489886, 0.003128626, -0.01...  \n",
       "21  [-0.03346974, -0.07286795, -0.13021637, -0.007...  \n",
       "22  [0.09801802, 0.020838303, 0.035793103, 0.05458...  \n",
       "23  [-0.071274094, -0.07724204, -0.054615665, -0.0...  \n",
       "24  [-0.06612152, -0.05528904, -0.0983323, 0.01714...  \n",
       "25  [-0.0019091254, 0.014995057, 0.04013634, -0.02...  \n",
       "26  [0.018412575, 0.014546668, -0.023422927, 0.094...  \n",
       "27  [0.042410787, -0.049514998, -0.061050948, 0.04...  \n",
       "28  [-0.017061016, -0.087328464, -0.019075712, 0.1...  \n",
       "29  [0.022794552, -0.11553865, 0.012081963, 0.0831...  \n",
       "30  [0.008736651, -0.0027261714, 0.015496009, 0.02...  \n",
       "31  [0.009260033, -0.036371656, 0.05298036, -0.055...  \n",
       "32  [0.002926987, 0.010527462, -0.03377514, -0.068...  \n",
       "33  [-0.039358903, 0.015897823, -0.035706297, -0.0...  \n",
       "34  [-0.028012691, -0.038667988, -0.022580912, -0....  \n",
       "35  [-0.009591965, -0.0054962244, -0.008626771, -0...  \n",
       "36  [-0.047342326, 0.03945248, 0.030565826, -0.049...  \n",
       "37  [0.03776277, 0.026553273, 0.022663444, -0.0016...  \n",
       "38  [0.040732816, 0.004375422, -0.02246262, -0.014...  \n",
       "39  [0.06212287, -0.119477965, 0.002100817, 0.0573...  \n",
       "40  [-0.035855744, -0.0048338864, -0.008274413, 0....  \n",
       "41  [-0.00273796, -0.013839726, 0.014343991, -0.03...  \n",
       "42  [-0.055298418, 0.01586196, -0.011672523, 0.104...  \n",
       "43  [-0.047459356, 0.041661527, -0.021046067, -0.0...  \n",
       "44  [-0.060417198, 0.0048286803, -0.03349639, 0.11...  \n",
       "45  [0.023872152, 0.008181557, 0.032098502, -0.001...  \n",
       "46  [0.042925302, -0.009820527, -0.02448054, -0.02...  \n",
       "47  [0.04361059, -0.060083244, 0.026317095, 0.0907...  \n",
       "48  [0.018182483, -0.08957952, 0.048921805, -0.066...  \n",
       "49  [-0.022645863, 0.0143042, -8.128404e-05, -0.06...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('filteredData80/mergedTables80.csv', nrows=50)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efa41980-f3cd-44bd-a700-1e5e66436999",
   "metadata": {},
   "source": [
    "### Data Aggregation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e49eab16-4532-4502-9328-c8c1018d5b9c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 'filteredData90/mergedTables90.csv' \n"
     ]
    }
   ],
   "source": [
    "filename = \"filteredData90/mergedTables90.csv\"\n",
    "grouping_column = \"genome_id\"\n",
    "\n",
    "\n",
    "total_sums = pd.DataFrame()\n",
    "total_counts = pd.DataFrame()\n",
    "\n",
    "df_iterator = pd.read_csv(filename, chunksize=5000, header=0)\n",
    "\n",
    "print(f\"Processing '{filename}' \")\n",
    "\n",
    "# Stores the running totals for each genome_id \n",
    "# Each genome_id maps to a list with [0] = numpy array sum, [1] = instance count\n",
    "embedding_aggregates = {}\n",
    "\n",
    "for chunk in df_iterator:\n",
    "\n",
    "    # converts 'embedding' column back into a vector\n",
    "    embeddings = chunk['embedding'].apply(ast.literal_eval)\n",
    "\n",
    "    \n",
    "    for genome_id, vector in zip(chunk[grouping_column], embeddings):\n",
    "\n",
    "        # Converts the embedding vector into a numpy array for faster addition\n",
    "        vector_np = np.array(vector, dtype=float)\n",
    "\n",
    "        if genome_id in embedding_aggregates:\n",
    "                # If genome_id exists, add to the existing vector and increment count\n",
    "                embedding_aggregates[genome_id][0] += vector_np\n",
    "                embedding_aggregates[genome_id][1] += 1\n",
    "        else:\n",
    "                # If it's a new genome_id, initialize it\n",
    "                embedding_aggregates[genome_id] = [vector_np, 1]\n",
    "\n",
    "# Calculate the final averages and store in the final means table\n",
    "final_means = {}\n",
    "for genome_id, (sum_vector, count) in embedding_aggregates.items():\n",
    "    final_means[genome_id] = sum_vector / count\n",
    "\n",
    "# Convert this dictionary to a dataframe for viewing\n",
    "mean_embeddings_df = pd.DataFrame.from_dict(final_means)\n",
    "\n",
    "df_transposed = mean_embeddings_df.T\n",
    "df_renamed = df_transposed.rename_axis('genome_id')\n",
    "\n",
    "df_renamed.to_csv('meanEmbeddings90.csv',mode='w', index=True)\n",
    "\n",
    "print(\"\\n--- Final Mean Values per GENOME ---\")\n",
    "print(mean_embeddings_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "564ee1e6-0c35-43f9-84d6-5c13fb3d3095",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"meanEmbeddings80.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ee990be2-a5bf-4301-946f-4d1e09ed234c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>1270</th>\n",
       "      <th>1271</th>\n",
       "      <th>1272</th>\n",
       "      <th>1273</th>\n",
       "      <th>1274</th>\n",
       "      <th>1275</th>\n",
       "      <th>1276</th>\n",
       "      <th>1277</th>\n",
       "      <th>1278</th>\n",
       "      <th>1279</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_009811375.1</td>\n",
       "      <td>0.005097</td>\n",
       "      <td>-0.029108</td>\n",
       "      <td>0.009158</td>\n",
       "      <td>-0.019778</td>\n",
       "      <td>-0.047010</td>\n",
       "      <td>0.012223</td>\n",
       "      <td>0.010697</td>\n",
       "      <td>-0.041409</td>\n",
       "      <td>0.018837</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017796</td>\n",
       "      <td>-0.053621</td>\n",
       "      <td>-0.030693</td>\n",
       "      <td>-0.001465</td>\n",
       "      <td>-0.020989</td>\n",
       "      <td>0.028310</td>\n",
       "      <td>0.048492</td>\n",
       "      <td>-0.133866</td>\n",
       "      <td>-0.008398</td>\n",
       "      <td>0.037826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_900112575.1</td>\n",
       "      <td>0.023043</td>\n",
       "      <td>0.010633</td>\n",
       "      <td>0.002235</td>\n",
       "      <td>0.074152</td>\n",
       "      <td>-0.033777</td>\n",
       "      <td>0.016179</td>\n",
       "      <td>-0.010999</td>\n",
       "      <td>-0.033452</td>\n",
       "      <td>0.019202</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074602</td>\n",
       "      <td>0.007044</td>\n",
       "      <td>-0.054413</td>\n",
       "      <td>0.024610</td>\n",
       "      <td>-0.012493</td>\n",
       "      <td>0.012762</td>\n",
       "      <td>0.052699</td>\n",
       "      <td>-0.044988</td>\n",
       "      <td>-0.076092</td>\n",
       "      <td>0.005212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_022701015.1</td>\n",
       "      <td>0.018318</td>\n",
       "      <td>-0.014619</td>\n",
       "      <td>-0.006187</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>-0.056363</td>\n",
       "      <td>-0.003505</td>\n",
       "      <td>0.011603</td>\n",
       "      <td>-0.034121</td>\n",
       "      <td>0.013086</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050423</td>\n",
       "      <td>-0.013335</td>\n",
       "      <td>-0.044080</td>\n",
       "      <td>0.023759</td>\n",
       "      <td>-0.022140</td>\n",
       "      <td>0.021541</td>\n",
       "      <td>0.049170</td>\n",
       "      <td>-0.094708</td>\n",
       "      <td>-0.053135</td>\n",
       "      <td>0.021463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_029223485.1</td>\n",
       "      <td>-0.006340</td>\n",
       "      <td>-0.001274</td>\n",
       "      <td>-0.011235</td>\n",
       "      <td>-0.016100</td>\n",
       "      <td>-0.035799</td>\n",
       "      <td>0.042902</td>\n",
       "      <td>-0.044398</td>\n",
       "      <td>-0.017327</td>\n",
       "      <td>0.016086</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043303</td>\n",
       "      <td>-0.023968</td>\n",
       "      <td>-0.038528</td>\n",
       "      <td>0.008071</td>\n",
       "      <td>-0.029596</td>\n",
       "      <td>0.045339</td>\n",
       "      <td>-0.017117</td>\n",
       "      <td>-0.095326</td>\n",
       "      <td>0.026001</td>\n",
       "      <td>0.049010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_039532985.1</td>\n",
       "      <td>-0.022140</td>\n",
       "      <td>-0.012035</td>\n",
       "      <td>-0.026132</td>\n",
       "      <td>-0.019179</td>\n",
       "      <td>-0.026588</td>\n",
       "      <td>0.061969</td>\n",
       "      <td>-0.039426</td>\n",
       "      <td>-0.031366</td>\n",
       "      <td>0.008159</td>\n",
       "      <td>...</td>\n",
       "      <td>0.037482</td>\n",
       "      <td>-0.046364</td>\n",
       "      <td>-0.051875</td>\n",
       "      <td>0.007019</td>\n",
       "      <td>-0.032136</td>\n",
       "      <td>0.034622</td>\n",
       "      <td>-0.028119</td>\n",
       "      <td>-0.112294</td>\n",
       "      <td>0.045929</td>\n",
       "      <td>0.050990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6593</th>\n",
       "      <td>GCF_008330085.1</td>\n",
       "      <td>0.049581</td>\n",
       "      <td>-0.025186</td>\n",
       "      <td>0.004193</td>\n",
       "      <td>-0.031469</td>\n",
       "      <td>0.024565</td>\n",
       "      <td>-0.041043</td>\n",
       "      <td>0.038940</td>\n",
       "      <td>-0.093084</td>\n",
       "      <td>0.076647</td>\n",
       "      <td>...</td>\n",
       "      <td>0.029283</td>\n",
       "      <td>0.028986</td>\n",
       "      <td>-0.038318</td>\n",
       "      <td>0.005205</td>\n",
       "      <td>0.025760</td>\n",
       "      <td>0.025818</td>\n",
       "      <td>0.094213</td>\n",
       "      <td>-0.128978</td>\n",
       "      <td>-0.060159</td>\n",
       "      <td>0.012950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6594</th>\n",
       "      <td>GCF_014287475.1</td>\n",
       "      <td>0.018989</td>\n",
       "      <td>0.014545</td>\n",
       "      <td>0.036779</td>\n",
       "      <td>0.095246</td>\n",
       "      <td>-0.085226</td>\n",
       "      <td>0.002204</td>\n",
       "      <td>0.020265</td>\n",
       "      <td>-0.037941</td>\n",
       "      <td>-0.046601</td>\n",
       "      <td>...</td>\n",
       "      <td>0.110950</td>\n",
       "      <td>-0.079281</td>\n",
       "      <td>-0.048010</td>\n",
       "      <td>0.020628</td>\n",
       "      <td>-0.041545</td>\n",
       "      <td>-0.039500</td>\n",
       "      <td>0.106955</td>\n",
       "      <td>-0.061335</td>\n",
       "      <td>-0.043488</td>\n",
       "      <td>0.063484</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6595</th>\n",
       "      <td>GCF_003096355.1</td>\n",
       "      <td>0.002901</td>\n",
       "      <td>-0.006817</td>\n",
       "      <td>0.060108</td>\n",
       "      <td>0.063684</td>\n",
       "      <td>-0.064929</td>\n",
       "      <td>0.028600</td>\n",
       "      <td>0.003512</td>\n",
       "      <td>-0.077551</td>\n",
       "      <td>-0.009452</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078974</td>\n",
       "      <td>-0.093890</td>\n",
       "      <td>-0.070994</td>\n",
       "      <td>-0.020510</td>\n",
       "      <td>-0.043842</td>\n",
       "      <td>0.006273</td>\n",
       "      <td>0.033756</td>\n",
       "      <td>-0.151653</td>\n",
       "      <td>0.039770</td>\n",
       "      <td>0.029558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6596</th>\n",
       "      <td>GCF_020320395.1</td>\n",
       "      <td>-0.030550</td>\n",
       "      <td>0.035787</td>\n",
       "      <td>0.044445</td>\n",
       "      <td>-0.035183</td>\n",
       "      <td>-0.020740</td>\n",
       "      <td>0.001623</td>\n",
       "      <td>-0.006783</td>\n",
       "      <td>0.005429</td>\n",
       "      <td>0.013815</td>\n",
       "      <td>...</td>\n",
       "      <td>0.039419</td>\n",
       "      <td>-0.008988</td>\n",
       "      <td>-0.058821</td>\n",
       "      <td>-0.004026</td>\n",
       "      <td>-0.040824</td>\n",
       "      <td>0.077711</td>\n",
       "      <td>0.076654</td>\n",
       "      <td>-0.073196</td>\n",
       "      <td>0.035232</td>\n",
       "      <td>0.021135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6597</th>\n",
       "      <td>GCF_003194565.2</td>\n",
       "      <td>-0.024084</td>\n",
       "      <td>-0.066134</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.027406</td>\n",
       "      <td>-0.113009</td>\n",
       "      <td>-0.038454</td>\n",
       "      <td>0.061566</td>\n",
       "      <td>-0.088110</td>\n",
       "      <td>0.012641</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014469</td>\n",
       "      <td>-0.044773</td>\n",
       "      <td>-0.082052</td>\n",
       "      <td>0.018116</td>\n",
       "      <td>-0.013950</td>\n",
       "      <td>-0.119604</td>\n",
       "      <td>-0.017044</td>\n",
       "      <td>-0.127545</td>\n",
       "      <td>0.029764</td>\n",
       "      <td>0.108800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6598 rows × 1281 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            genome_id         0         1         2         3         4  \\\n",
       "0     GCF_009811375.1  0.005097 -0.029108  0.009158 -0.019778 -0.047010   \n",
       "1     GCF_900112575.1  0.023043  0.010633  0.002235  0.074152 -0.033777   \n",
       "2     GCF_022701015.1  0.018318 -0.014619 -0.006187  0.018433 -0.056363   \n",
       "3     GCF_029223485.1 -0.006340 -0.001274 -0.011235 -0.016100 -0.035799   \n",
       "4     GCF_039532985.1 -0.022140 -0.012035 -0.026132 -0.019179 -0.026588   \n",
       "...               ...       ...       ...       ...       ...       ...   \n",
       "6593  GCF_008330085.1  0.049581 -0.025186  0.004193 -0.031469  0.024565   \n",
       "6594  GCF_014287475.1  0.018989  0.014545  0.036779  0.095246 -0.085226   \n",
       "6595  GCF_003096355.1  0.002901 -0.006817  0.060108  0.063684 -0.064929   \n",
       "6596  GCF_020320395.1 -0.030550  0.035787  0.044445 -0.035183 -0.020740   \n",
       "6597  GCF_003194565.2 -0.024084 -0.066134  0.015158  0.027406 -0.113009   \n",
       "\n",
       "             5         6         7         8  ...      1270      1271  \\\n",
       "0     0.012223  0.010697 -0.041409  0.018837  ...  0.017796 -0.053621   \n",
       "1     0.016179 -0.010999 -0.033452  0.019202  ...  0.074602  0.007044   \n",
       "2    -0.003505  0.011603 -0.034121  0.013086  ...  0.050423 -0.013335   \n",
       "3     0.042902 -0.044398 -0.017327  0.016086  ...  0.043303 -0.023968   \n",
       "4     0.061969 -0.039426 -0.031366  0.008159  ...  0.037482 -0.046364   \n",
       "...        ...       ...       ...       ...  ...       ...       ...   \n",
       "6593 -0.041043  0.038940 -0.093084  0.076647  ...  0.029283  0.028986   \n",
       "6594  0.002204  0.020265 -0.037941 -0.046601  ...  0.110950 -0.079281   \n",
       "6595  0.028600  0.003512 -0.077551 -0.009452  ...  0.078974 -0.093890   \n",
       "6596  0.001623 -0.006783  0.005429  0.013815  ...  0.039419 -0.008988   \n",
       "6597 -0.038454  0.061566 -0.088110  0.012641  ...  0.014469 -0.044773   \n",
       "\n",
       "          1272      1273      1274      1275      1276      1277      1278  \\\n",
       "0    -0.030693 -0.001465 -0.020989  0.028310  0.048492 -0.133866 -0.008398   \n",
       "1    -0.054413  0.024610 -0.012493  0.012762  0.052699 -0.044988 -0.076092   \n",
       "2    -0.044080  0.023759 -0.022140  0.021541  0.049170 -0.094708 -0.053135   \n",
       "3    -0.038528  0.008071 -0.029596  0.045339 -0.017117 -0.095326  0.026001   \n",
       "4    -0.051875  0.007019 -0.032136  0.034622 -0.028119 -0.112294  0.045929   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "6593 -0.038318  0.005205  0.025760  0.025818  0.094213 -0.128978 -0.060159   \n",
       "6594 -0.048010  0.020628 -0.041545 -0.039500  0.106955 -0.061335 -0.043488   \n",
       "6595 -0.070994 -0.020510 -0.043842  0.006273  0.033756 -0.151653  0.039770   \n",
       "6596 -0.058821 -0.004026 -0.040824  0.077711  0.076654 -0.073196  0.035232   \n",
       "6597 -0.082052  0.018116 -0.013950 -0.119604 -0.017044 -0.127545  0.029764   \n",
       "\n",
       "          1279  \n",
       "0     0.037826  \n",
       "1     0.005212  \n",
       "2     0.021463  \n",
       "3     0.049010  \n",
       "4     0.050990  \n",
       "...        ...  \n",
       "6593  0.012950  \n",
       "6594  0.063484  \n",
       "6595  0.029558  \n",
       "6596  0.021135  \n",
       "6597  0.108800  \n",
       "\n",
       "[6598 rows x 1281 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9e236ce4-0b1f-4222-8839-37a999d1a6fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>...</th>\n",
       "      <th>1270</th>\n",
       "      <th>1271</th>\n",
       "      <th>1272</th>\n",
       "      <th>1273</th>\n",
       "      <th>1274</th>\n",
       "      <th>1275</th>\n",
       "      <th>1276</th>\n",
       "      <th>1277</th>\n",
       "      <th>1278</th>\n",
       "      <th>1279</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCF_009811375.1</td>\n",
       "      <td>0.005097</td>\n",
       "      <td>-0.029108</td>\n",
       "      <td>0.009158</td>\n",
       "      <td>-0.019778</td>\n",
       "      <td>-0.047010</td>\n",
       "      <td>0.012223</td>\n",
       "      <td>0.010697</td>\n",
       "      <td>-0.041409</td>\n",
       "      <td>0.018837</td>\n",
       "      <td>...</td>\n",
       "      <td>0.017796</td>\n",
       "      <td>-0.053621</td>\n",
       "      <td>-0.030693</td>\n",
       "      <td>-0.001465</td>\n",
       "      <td>-0.020989</td>\n",
       "      <td>0.028310</td>\n",
       "      <td>0.048492</td>\n",
       "      <td>-0.133866</td>\n",
       "      <td>-0.008398</td>\n",
       "      <td>0.037826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCF_900112575.1</td>\n",
       "      <td>0.023043</td>\n",
       "      <td>0.010633</td>\n",
       "      <td>0.002235</td>\n",
       "      <td>0.074152</td>\n",
       "      <td>-0.033777</td>\n",
       "      <td>0.016179</td>\n",
       "      <td>-0.010999</td>\n",
       "      <td>-0.033452</td>\n",
       "      <td>0.019202</td>\n",
       "      <td>...</td>\n",
       "      <td>0.074602</td>\n",
       "      <td>0.007044</td>\n",
       "      <td>-0.054413</td>\n",
       "      <td>0.024610</td>\n",
       "      <td>-0.012493</td>\n",
       "      <td>0.012762</td>\n",
       "      <td>0.052699</td>\n",
       "      <td>-0.044988</td>\n",
       "      <td>-0.076092</td>\n",
       "      <td>0.005212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCF_022701015.1</td>\n",
       "      <td>0.018318</td>\n",
       "      <td>-0.014619</td>\n",
       "      <td>-0.006187</td>\n",
       "      <td>0.018433</td>\n",
       "      <td>-0.056363</td>\n",
       "      <td>-0.003505</td>\n",
       "      <td>0.011603</td>\n",
       "      <td>-0.034121</td>\n",
       "      <td>0.013086</td>\n",
       "      <td>...</td>\n",
       "      <td>0.050423</td>\n",
       "      <td>-0.013335</td>\n",
       "      <td>-0.044080</td>\n",
       "      <td>0.023759</td>\n",
       "      <td>-0.022140</td>\n",
       "      <td>0.021541</td>\n",
       "      <td>0.049170</td>\n",
       "      <td>-0.094708</td>\n",
       "      <td>-0.053135</td>\n",
       "      <td>0.021463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCF_029223485.1</td>\n",
       "      <td>-0.006340</td>\n",
       "      <td>-0.001274</td>\n",
       "      <td>-0.011235</td>\n",
       "      <td>-0.016100</td>\n",
       "      <td>-0.035799</td>\n",
       "      <td>0.042902</td>\n",
       "      <td>-0.044398</td>\n",
       "      <td>-0.017327</td>\n",
       "      <td>0.016086</td>\n",
       "      <td>...</td>\n",
       "      <td>0.043303</td>\n",
       "      <td>-0.023968</td>\n",
       "      <td>-0.038528</td>\n",
       "      <td>0.008071</td>\n",
       "      <td>-0.029596</td>\n",
       "      <td>0.045339</td>\n",
       "      <td>-0.017117</td>\n",
       "      <td>-0.095326</td>\n",
       "      <td>0.026001</td>\n",
       "      <td>0.049010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCF_039532985.1</td>\n",
       "      <td>-0.022140</td>\n",
       "      <td>-0.012035</td>\n",
       "      <td>-0.026132</td>\n",
       "      <td>-0.019179</td>\n",
       "      <td>-0.026588</td>\n",
       "      <td>0.061969</td>\n",
       "      <td>-0.039426</td>\n",
       "      <td>-0.031366</td>\n",
       "      <td>0.008159</td>\n",
       "      <td>...</td>\n",
       "      <td>0.037482</td>\n",
       "      <td>-0.046364</td>\n",
       "      <td>-0.051875</td>\n",
       "      <td>0.007019</td>\n",
       "      <td>-0.032136</td>\n",
       "      <td>0.034622</td>\n",
       "      <td>-0.028119</td>\n",
       "      <td>-0.112294</td>\n",
       "      <td>0.045929</td>\n",
       "      <td>0.050990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6778</th>\n",
       "      <td>GCF_003096355.1</td>\n",
       "      <td>0.002901</td>\n",
       "      <td>-0.006817</td>\n",
       "      <td>0.060108</td>\n",
       "      <td>0.063684</td>\n",
       "      <td>-0.064929</td>\n",
       "      <td>0.028600</td>\n",
       "      <td>0.003512</td>\n",
       "      <td>-0.077551</td>\n",
       "      <td>-0.009452</td>\n",
       "      <td>...</td>\n",
       "      <td>0.078974</td>\n",
       "      <td>-0.093890</td>\n",
       "      <td>-0.070994</td>\n",
       "      <td>-0.020510</td>\n",
       "      <td>-0.043842</td>\n",
       "      <td>0.006273</td>\n",
       "      <td>0.033756</td>\n",
       "      <td>-0.151653</td>\n",
       "      <td>0.039770</td>\n",
       "      <td>0.029558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6779</th>\n",
       "      <td>GCF_042654745.1</td>\n",
       "      <td>0.003836</td>\n",
       "      <td>-0.058023</td>\n",
       "      <td>0.018270</td>\n",
       "      <td>-0.060214</td>\n",
       "      <td>-0.057618</td>\n",
       "      <td>-0.008462</td>\n",
       "      <td>0.006931</td>\n",
       "      <td>-0.140456</td>\n",
       "      <td>0.025044</td>\n",
       "      <td>...</td>\n",
       "      <td>0.041891</td>\n",
       "      <td>0.038329</td>\n",
       "      <td>0.007000</td>\n",
       "      <td>-0.032080</td>\n",
       "      <td>-0.049002</td>\n",
       "      <td>0.039142</td>\n",
       "      <td>0.036184</td>\n",
       "      <td>-0.132186</td>\n",
       "      <td>-0.030021</td>\n",
       "      <td>0.089851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6780</th>\n",
       "      <td>GCF_020320395.1</td>\n",
       "      <td>-0.030550</td>\n",
       "      <td>0.035787</td>\n",
       "      <td>0.044445</td>\n",
       "      <td>-0.035183</td>\n",
       "      <td>-0.020740</td>\n",
       "      <td>0.001623</td>\n",
       "      <td>-0.006783</td>\n",
       "      <td>0.005429</td>\n",
       "      <td>0.013815</td>\n",
       "      <td>...</td>\n",
       "      <td>0.039419</td>\n",
       "      <td>-0.008988</td>\n",
       "      <td>-0.058821</td>\n",
       "      <td>-0.004026</td>\n",
       "      <td>-0.040824</td>\n",
       "      <td>0.077711</td>\n",
       "      <td>0.076654</td>\n",
       "      <td>-0.073196</td>\n",
       "      <td>0.035232</td>\n",
       "      <td>0.021135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6781</th>\n",
       "      <td>GCF_900104245.1</td>\n",
       "      <td>-0.035257</td>\n",
       "      <td>0.044724</td>\n",
       "      <td>-0.031666</td>\n",
       "      <td>0.011155</td>\n",
       "      <td>0.052581</td>\n",
       "      <td>0.044551</td>\n",
       "      <td>-0.056487</td>\n",
       "      <td>0.077214</td>\n",
       "      <td>0.023585</td>\n",
       "      <td>...</td>\n",
       "      <td>0.099217</td>\n",
       "      <td>-0.060239</td>\n",
       "      <td>-0.015377</td>\n",
       "      <td>0.043792</td>\n",
       "      <td>0.009540</td>\n",
       "      <td>0.013731</td>\n",
       "      <td>0.059342</td>\n",
       "      <td>-0.104578</td>\n",
       "      <td>0.025894</td>\n",
       "      <td>0.020628</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6782</th>\n",
       "      <td>GCF_003194565.2</td>\n",
       "      <td>-0.024084</td>\n",
       "      <td>-0.066134</td>\n",
       "      <td>0.015158</td>\n",
       "      <td>0.027406</td>\n",
       "      <td>-0.113009</td>\n",
       "      <td>-0.038454</td>\n",
       "      <td>0.061566</td>\n",
       "      <td>-0.088110</td>\n",
       "      <td>0.012641</td>\n",
       "      <td>...</td>\n",
       "      <td>0.014469</td>\n",
       "      <td>-0.044773</td>\n",
       "      <td>-0.082052</td>\n",
       "      <td>0.018116</td>\n",
       "      <td>-0.013950</td>\n",
       "      <td>-0.119604</td>\n",
       "      <td>-0.017044</td>\n",
       "      <td>-0.127545</td>\n",
       "      <td>0.029764</td>\n",
       "      <td>0.108800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6783 rows × 1281 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            genome_id         0         1         2         3         4  \\\n",
       "0     GCF_009811375.1  0.005097 -0.029108  0.009158 -0.019778 -0.047010   \n",
       "1     GCF_900112575.1  0.023043  0.010633  0.002235  0.074152 -0.033777   \n",
       "2     GCF_022701015.1  0.018318 -0.014619 -0.006187  0.018433 -0.056363   \n",
       "3     GCF_029223485.1 -0.006340 -0.001274 -0.011235 -0.016100 -0.035799   \n",
       "4     GCF_039532985.1 -0.022140 -0.012035 -0.026132 -0.019179 -0.026588   \n",
       "...               ...       ...       ...       ...       ...       ...   \n",
       "6778  GCF_003096355.1  0.002901 -0.006817  0.060108  0.063684 -0.064929   \n",
       "6779  GCF_042654745.1  0.003836 -0.058023  0.018270 -0.060214 -0.057618   \n",
       "6780  GCF_020320395.1 -0.030550  0.035787  0.044445 -0.035183 -0.020740   \n",
       "6781  GCF_900104245.1 -0.035257  0.044724 -0.031666  0.011155  0.052581   \n",
       "6782  GCF_003194565.2 -0.024084 -0.066134  0.015158  0.027406 -0.113009   \n",
       "\n",
       "             5         6         7         8  ...      1270      1271  \\\n",
       "0     0.012223  0.010697 -0.041409  0.018837  ...  0.017796 -0.053621   \n",
       "1     0.016179 -0.010999 -0.033452  0.019202  ...  0.074602  0.007044   \n",
       "2    -0.003505  0.011603 -0.034121  0.013086  ...  0.050423 -0.013335   \n",
       "3     0.042902 -0.044398 -0.017327  0.016086  ...  0.043303 -0.023968   \n",
       "4     0.061969 -0.039426 -0.031366  0.008159  ...  0.037482 -0.046364   \n",
       "...        ...       ...       ...       ...  ...       ...       ...   \n",
       "6778  0.028600  0.003512 -0.077551 -0.009452  ...  0.078974 -0.093890   \n",
       "6779 -0.008462  0.006931 -0.140456  0.025044  ...  0.041891  0.038329   \n",
       "6780  0.001623 -0.006783  0.005429  0.013815  ...  0.039419 -0.008988   \n",
       "6781  0.044551 -0.056487  0.077214  0.023585  ...  0.099217 -0.060239   \n",
       "6782 -0.038454  0.061566 -0.088110  0.012641  ...  0.014469 -0.044773   \n",
       "\n",
       "          1272      1273      1274      1275      1276      1277      1278  \\\n",
       "0    -0.030693 -0.001465 -0.020989  0.028310  0.048492 -0.133866 -0.008398   \n",
       "1    -0.054413  0.024610 -0.012493  0.012762  0.052699 -0.044988 -0.076092   \n",
       "2    -0.044080  0.023759 -0.022140  0.021541  0.049170 -0.094708 -0.053135   \n",
       "3    -0.038528  0.008071 -0.029596  0.045339 -0.017117 -0.095326  0.026001   \n",
       "4    -0.051875  0.007019 -0.032136  0.034622 -0.028119 -0.112294  0.045929   \n",
       "...        ...       ...       ...       ...       ...       ...       ...   \n",
       "6778 -0.070994 -0.020510 -0.043842  0.006273  0.033756 -0.151653  0.039770   \n",
       "6779  0.007000 -0.032080 -0.049002  0.039142  0.036184 -0.132186 -0.030021   \n",
       "6780 -0.058821 -0.004026 -0.040824  0.077711  0.076654 -0.073196  0.035232   \n",
       "6781 -0.015377  0.043792  0.009540  0.013731  0.059342 -0.104578  0.025894   \n",
       "6782 -0.082052  0.018116 -0.013950 -0.119604 -0.017044 -0.127545  0.029764   \n",
       "\n",
       "          1279  \n",
       "0     0.037826  \n",
       "1     0.005212  \n",
       "2     0.021463  \n",
       "3     0.049010  \n",
       "4     0.050990  \n",
       "...        ...  \n",
       "6778  0.029558  \n",
       "6779  0.089851  \n",
       "6780  0.021135  \n",
       "6781  0.020628  \n",
       "6782  0.108800  \n",
       "\n",
       "[6783 rows x 1281 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test = pd.read_csv(\"meanEmbeddings80.csv\")\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f866a04-c171-4c40-bdd0-f6fea6e0d5fc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (bacteria-env)",
   "language": "python",
   "name": "bacteria-env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
